ops/framework.py

```diff
@@ -631,6 +631,22 @@ def __init__(self, storage: Union[SQLiteStorage, JujuStorage], charm_dir: 'Path'
         self._juju_debug_at = (set(x.strip() for x in debug_at.split(','))
                                if debug_at else set())  # type: Set[str]
 
+    def _clear_all_state(self):
+        """Clear the state.
+
+        Allows testing environments to clean the framework up and instantiate the
+        same Object (Handle, specifically) multiple times.
+        """
+        if self.model._backend._hook_is_running:  # type: ignore # noqa
+            raise RuntimeError('You should not do this while a hook is running. '
+                               'This method is for testing only.')
+
+        self._objects.clear()
+        self._observers.clear()
+        self._observer.clear()
+        self._type_registry.clear()
+        self._type_known.clear()
```

Comment:
FWIW, I don't think this is actually enough. AIUI the issue is that the Python code itself is actually adding bound methods to classes that point to this framework. So if we really wanted this, we would need to go around and remove all of those methods. (I'm not 100% sure what else, but there are a lot of Global modifications that are being done during initialization and registration of classes.)